dm/syncer/validator_checkpoint.go

```diff
@@ -0,0 +1,436 @@
+
+	var location *binlog.Location
+
+	// at most one row
+	if rows.Next() {
+		var (
+			binlogName, binlogGtidStr string
+			binlogPos                 uint32
+		)
+
+		err = rows.Scan(&binlogName, &binlogPos, &binlogGtidStr)
+		if err != nil {
+			return nil, err
+		}
+		gset, err2 := gtid.ParserGTID(c.cfg.Flavor, binlogGtidStr)
+		if err2 != nil {
+			return nil, err2
+		}
+		tmpLoc := binlog.InitLocation(mysql.Position{Name: binlogName, Pos: binlogPos}, gset)
+		location = &tmpLoc
+	}
+	if err = rows.Err(); err != nil {
+		return nil, err
+	}
+	return location, nil
+}
+
+func (c *validatorPersistHelper) loadPendingChange(tctx *tcontext.Context) (map[string]*tableChange, int64, error) {
+	res := make(map[string]*tableChange)
+	rev := int64(1)
+	sql := "select schema_name, table_name, row_pk, data, revision from " + c.pendingChangeTableName + " where source = ?"
+	rows, err := c.dbConn.QuerySQL(tctx, sql, c.cfg.SourceID)
+	if err != nil {
+		return nil, 0, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var (
+			schemaName, tableName, key string
+			data                       []byte
+			revision                   int64
+		)
+		err = rows.Scan(&schemaName, &tableName, &key, &data, &revision)
+		if err != nil {
+			return nil, 0, err
+		}
+		sourceTbl := filter.Table{Schema: schemaName, Name: tableName}
+		fullTableName := sourceTbl.String()
+		tblChange, ok := res[fullTableName]
+		if !ok {
+			tblChange = newTableChange(&validateTableInfo{Source: &sourceTbl})
+			res[fullTableName] = tblChange
+		}
+		var row *rowChange
+		err = json.Unmarshal(data, &row)
+		if err != nil {
+			return nil, 0, err
+		}
+		tblChange.rows[key] = row
+		rev = revision
+	}
+
+	if err = rows.Err(); err != nil {
+		return nil, 0, err
+	}
+	return res, rev, nil
+}
+
+func (c *validatorPersistHelper) loadTableStatus(tctx *tcontext.Context) (map[string]*tableValidateStatus, error) {
+	res := make(map[string]*tableValidateStatus)
+	sql := "select src_schema_name, src_table_name, dst_schema_name, dst_table_name, stage, message from " +
+		c.tableStatusTableName + " where source = ?"
+	rows, err := c.dbConn.QuerySQL(tctx, sql, c.cfg.SourceID)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var (
+			srcSchemaName, srcTableName, dstSchemaName, dstTableName string
+			stage                                                    int
+			message                                                  string
+		)
+		err = rows.Scan(&srcSchemaName, &srcTableName, &dstSchemaName, &dstTableName, &stage, &message)
+		if err != nil {
+			return nil, err
+		}
+		srcTbl := filter.Table{Schema: srcSchemaName, Name: srcTableName}
+		fullTableName := srcTbl.String()
+		res[fullTableName] = &tableValidateStatus{
+			source:  srcTbl,
+			target:  filter.Table{Schema: dstSchemaName, Name: dstTableName},
+			stage:   pb.Stage(stage),
+			message: message,
+		}
+	}
+
+	if err = rows.Err(); err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
```

Comment:
What about adding a method to get `validation error` like `loadError`?

Reply:
add a load error count in later pr, loadError's not needed in this pr